sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -1741,7 +1743,8 @@ class Dataset[T] private[sql](
   def union(other: Dataset[T]): Dataset[T] = withSetOperator {
     // This breaks caching, but it's usually ok because it addresses a very specific use case:
     // using union to union many files or partitions.
-    CombineUnions(Union(logicalPlan, other.logicalPlan))
+    CombineUnions(Union(EliminateBarriers(logicalPlan), EliminateBarriers(other.logicalPlan)))
+      .mapChildren(AnalysisBarrier)
```

Comment:
This looks a bit ugly. But the barrier will prevent `CombineUnions` work normally.